Lib/ast.py

```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
         with self.block(extra=self.get_type_comment(node)):
             self.traverse(node.body)
 
+    def _str_literal_helper(
+        self, value, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimising escapes.
+        Returns (possible quote types, string literal to write).
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
+        val = "".join(
+            c.encode('unicode_escape').decode('ascii')
+            if c in escape or (not c.isprintable() and c not in '\n\t') else c
+            for c in value
+        )
+        qts = [quote for quote in quote_types if quote not in val]
+        if "\n" in val:
+            qts = [quote for quote in qts if quote in ('"""', "'''")]
+        if not qts:
+            # If there aren't any possible quote_types, fallback to using repr
+            # on the original value. Try to use a quote_type from quote_types.
+            value = repr(value)
+            quote_type = next((q for q in quote_types if value[0] in q), value[0])
+            return value[1:-1], [quote_type]
+        if val:
+            # Sort so that we prefer '''"''' over """\""""
+            qts.sort(key=lambda q: q[0] == val[-1])
+            # If we're using triple quotes and we'd need to escape a final quote, escape
+            if qts[0][0] == val[-1]:
+                assert len(qts[0]) == 3
```

Comment:
Replace this assert with an if, and if the result is false call fallback. We dont want assertion errors to pop out if we can traverse that string

Reply:
The assert will always pass (as long as people pass valid quotes to quote_types, which is okay to assume since this is an internal helper). This is because we'd have filtered out the single quote that we need to escape already, since it's clearly in the string.

If this is confusing, I can remove the assert (although I like it, since it helps clarify the comment) — but we don't need to fallback for a condition that will never trigger.